dns: support cares_search for resolution

SirR4T opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)


Please note: this is my first time writing C++ code, so I'm not sure if this is what was required.
Current test cases are passing, though, but haven't added any tests for `cares_search` resolution.

Fixes: #17850 (based on approach suggested in #19548)

addaleax commented on src/cares_wrap.cc:
```diff
@@ -466,7 +466,11 @@ void cares_wrap_hostent_cpy(struct hostent* dest, struct hostent* src) {
 }
 
-class QueryWrap;
+template <void (*F)(ares_channel, const char *, int, int, ares_callback, void *)>
```
Style nit: Left-leaning pointers (e.g. `const char*`, not `const char *`, here and below)

If you’re unsure about something, we have a [C++ style guide](https://github.com/nodejs/node/blob/master/CPP_STYLE_GUIDE.md), and, very recently, introduced `make format-cpp CLANG_FORMAT_START=upstream/master` (you may need to run `make format-cpp-build` first)

SirR4T replied:
@addaleax thanks for the tip! this produces `tools/clang-format/node_modules/` though, in my working directory. Good idea to add this to `.gitignore` in this PR? or should that be a different PR by itself?

addaleax commented on src/cares_wrap.cc:
```diff
@@ -593,10 +597,10 @@ void ChannelWrap::EnsureServers() {
-
-class QueryWrap : public AsyncWrap {
+template <void (*F)(ares_channel, const char *, int, int, ares_callback, void *)>
+class BaseWrap : public AsyncWrap {
```
I’m wondering if it makes sense to do templating on this level, since it introduces a new parallel class hierarchy … maybe we could store the function as a member of `QueryWrap` instead, and pass it to the `QueryWrap` constructor?

That would also allow us to avoid the `this->…` oddities below…

SirR4T replied:
@addaleax not too sure how this works in C++ land. Could you throw some more light? For instance, If I pass in the cares function (say `ares_search`) to the base class (say `BaseWrap`), could the result still be a class (instead of what I assume will be an object)?

I think that was the advantage of templating, so I need not touch wherever `QueryWrap` is being used to create other objects. Maybe some code snippets would help me understand better.

addaleax replied:
@SirR4T 

> For instance, If I pass in the cares function (say `ares_search`) to the base class (say `BaseWrap`), could the result still be a class (instead of what I assume will be an object)?

I don’t really understand this, sorry … What do you mean by “pass to the base class”?

> I think that was the advantage of templating, so I need not touch wherever `QueryWrap` is being used to create other objects.

Doing this also duplicates all generated code and 

Basically, what you have is something like this:

```c++
template <int (*F)()>
class Base {
  /* ... */
};

template <typename BaseWrap>
class SubclassWrap1 : BaseWrap {
  /* ... */
};

/* ... */

template <typename Wrap>
void Query(/* ... */) {
  new Wrap(/*...*/);
}

// And then use:
Query<SubclassWrap1<Base<function1>>>();
Query<SubclassWrap1<Base<function2>>>();
Query<SubclassWrap2<Base<function1>>>();
Query<SubclassWrap2<Base<function2>>>();
/* ... */
```

While this works, it seems very complex in terms of the generated class hiearchy.

What I’m thinking does not modify the existing class hierachy, but rather passes along one more argument:

```c++
class Base {
  Base(int (*fn)()) : search_function_(fn) {}
  /* ... */
 private:
  /* ... */
  int (*search_function_)();
};

class SubclassWrap1 : Base {
  SubclassWrap1(..., int (*fn)()) : Base(..., fn) {}
  // Or, alternatively, we could switch to making the constructor forward 
  // all arguments regardless of what they actually are:
  template <typename... Args>
  SubclassWrap(Args&&... args) : Base(std::forward<Args>(args)...) {}
};

template <typename Wrap>
void Query(/* ... */) {
  new Wrap(..., use_function_1 ? function1 : function2);
}
```

(You could of course use a typedef instead of always spelling out the function type here)

vsemozhetbyt commented on doc/api/dns.md:
```diff
@@ -287,6 +287,10 @@ changes:
+  - `search` {boolean} Resolve the hostname using the local domain name or the
+    search list for host-name lookup.
+    When `true`, the resolver will use the search list, which can create extra
+    dns queries. **Default:** `false`.
```
Nit: dns -> DNS?

vsemozhetbyt commented on doc/api/dns.md:
```diff
@@ -311,6 +315,10 @@ changes:
+  - `search` {boolean} Resolve the hostname using the local domain name or the
+    search list for host-name lookup.
+    When `true`, the resolver will use the search list, which can create extra
+    dns queries. **Default:** `false`.
```
Ditto.

SirR4T commented:
@Trott : fixed and added new test cases in the same file. Am still not sure, though, that these would cover the `ares_search` functionality fully. If someone could guide me what kind of tests should be required specifically for search functionality, would be happy to add them.

silverwind commented:
Is there a reason why only `A` and `AAAA` query types would get this feature?

I think we should add a `options` object to `resolve` so all query types can benefit. There was a recent request to have search domain support specifically on `SRV` queries.

addaleax commented on src/cares_wrap.cc:
```diff
@@ -637,7 +649,7 @@ class QueryWrap : public AsyncWrap {
     auto data = static_cast<struct CaresAsyncData*>(handle->data);
 
-    QueryWrap* wrap = data->wrap;
+    BaseWrap* wrap = data->wrap;
```
@SirR4T It’s hard to tell until this PR is at the point where the option starts to work.

I’d personally not split `BaseWrap` into two `QueryWrap` and `SearchWrap`, and instead pass the new constructor parameter along from the subclasses, e.g.:

```c++

class QueryAnyWrap: public QueryWrap {
 public:
  QueryAnyWrap(ChannelWrap* channel, Local<Object> req_wrap_obj)
    : QueryWrap(channel, req_wrap_obj) {
  }
```

becomes

```c++

class QueryAnyWrap: public QueryWrap {
 public:
  QueryAnyWrap(ChannelWrap* channel, Local<Object> req_wrap_obj, ares_function* fn)
    : QueryWrap(channel, req_wrap_obj, fn) {
  }
```

and then decide which function to pass in at the call site (the `Query` function). But yeah, it’s hard to tell in the current state of things…

SirR4T replied:
@addaleax should 

```
  env->SetProtoMethod(channel_wrap, "queryAny", Query<QueryAnyWrap>);
  env->SetProtoMethod(channel_wrap, "queryA", Query<QueryAWrap>);
  env->SetProtoMethod(channel_wrap, "queryAaaa", Query<QueryAaaaWrap>);
  env->SetProtoMethod(channel_wrap, "queryCname", Query<QueryCnameWrap>);
  env->SetProtoMethod(channel_wrap, "queryMx", Query<QueryMxWrap>);
  env->SetProtoMethod(channel_wrap, "queryNs", Query<QueryNsWrap>);
  env->SetProtoMethod(channel_wrap, "queryTxt", Query<QueryTxtWrap>);
  env->SetProtoMethod(channel_wrap, "querySrv", Query<QuerySrvWrap>);
  env->SetProtoMethod(channel_wrap, "queryPtr", Query<QueryPtrWrap>);
  env->SetProtoMethod(channel_wrap, "queryNaptr", Query<QueryNaptrWrap>);
  env->SetProtoMethod(channel_wrap, "querySoa", Query<QuerySoaWrap>);
  env->SetProtoMethod(channel_wrap, "getHostByAddr", Query<GetHostByAddrWrap>);
```

become something like this:
```
  env->SetProtoMethod(channel_wrap, "queryAny", Query<QueryAnyWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "queryA", Query<QueryAWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "queryAaaa", Query<QueryAaaaWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "queryCname", Query<QueryCnameWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "queryMx", Query<QueryMxWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "queryNs", Query<QueryNsWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "queryTxt", Query<QueryTxtWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "querySrv", Query<QuerySrvWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "queryPtr", Query<QueryPtrWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "queryNaptr", Query<QueryNaptrWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "querySoa", Query<QuerySoaWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "getHostByAddr", Query<GetHostByAddrWrap, ares_query>);

  env->SetProtoMethod(channel_wrap, "searchAny", Query<QueryAnyWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchA", Query<QueryAWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchAaaa", Query<QueryAaaaWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchCname", Query<QueryCnameWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchMx", Query<QueryMxWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchNs", Query<QueryNsWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchTxt", Query<QueryTxtWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchSrv", Query<QuerySrvWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchPtr", Query<QueryPtrWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchNaptr", Query<QueryNaptrWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchSoa", Query<QuerySoaWrap, ares_search>);
```
?

Although that kind of seems to defeat the purpose of passing in a `search: true` option to C++ land.

At this point, I think I need some pointers on how to pass the `search: true` option into (and parse for, while in) C++ land.

addaleax replied:
@SirR4T Yes, that’s what I want to avoid… The suggestion from my comment above would allow you to inspect the `search` option inside the `Query<…>` function, and then pass `ares_search` or `ares_query` depending on whether it is set or not

SirR4T replied:
yay 🎉 

build passed, as well as js tests passed, and removed `SearchWrap` as well.

SirR4T commented:
@silverwind : working on this further, now passing the search option through. Do expect to see more changes, though, as i'm still a noob in C++.

addaleax commented on src/cares_wrap.cc:
```diff
@@ -743,8 +753,14 @@ class QueryWrap : public AsyncWrap {
+ public:
+  QueryWrap(ChannelWrap* channel, Local<Object> req_wrap_obj, ares_function* fn)
+      : BaseWrap(channel, req_wrap_obj, fn) {}
+};
```
I think we don’t need the `BaseWrap` layer of indirection anymore, right?

SirR4T replied:
obvious, once you point it out 😁

removed. now how do we provide search domains, so that we can write some meaningful tests for `search: true`?

addaleax replied:
@SirR4T I don’t know the feature well enough (yet?) to understand how to write good tests for it, but yes, I think we’re at that point generally. The [mock DNS module](https://github.com/nodejs/node/blob/master/test/common/README.md#dns-module) might help a bit?

SirR4T replied:
Seems to me that we will need to get the `ares_channel channel->ndomains` setup properly, for `ares_search()` to be meaningful (not default to using `ares_query()` instead).

As far as i can gather, there are [four ways of doing this](https://github.com/nodejs/node/blob/master/deps/cares/src/ares_init.c#L191-L211). Not sure which would be easiest to mock / replicate.

SirR4T replied:
strike that, i was going down the wrong rabbit hole.

Here's what's working, now: 

**Search for "en" fails, before any changes to `/etc/resolv.conf`:**
```
> dns.resolve4('en', {ttl: true, search: true}, (err, result) => { if (err) console.error(err); console.log(result); })
QueryReqWrap {
  bindingName: 'queryA',
  callback: [Function],
  hostname: 'en',
  oncomplete: [Function: onresolve],
  ttl: true,
  search: true,
  domain:
   Domain {
     domain: null,
     _events:
      { removeListener: [Function: updateExceptionCapture],
        newListener: [Function: updateExceptionCapture],
        error: [Function: debugDomainError] },
     _eventsCount: 3,
     _maxListeners: undefined,
     members: [] },
  channel:
   ChannelWrap {
     domain:
      Domain {
        domain: null,
        _events: [Object],
        _eventsCount: 3,
        _maxListeners: undefined,
        members: [] } } }
> { Error: queryA ESERVFAIL en
    at QueryReqWrap.onresolve [as oncomplete] (dns.js:197:19)
  errno: 'ESERVFAIL',
  code: 'ESERVFAIL',
  syscall: 'queryA',
  hostname: 'en' }
```
Same with `nslookup`:
```
$ nslookup "en"
;; Got SERVFAIL reply from 10.11.1.20, trying next server
Server:		10.11.1.19
Address:	10.11.1.19#53

** server can't find en: SERVFAIL

```

**After adding `search wikipedia.org` to `/etc/resolv.conf`:**
```
> dns.resolve4('en', {ttl: true, search: true}, (err, result) => { if (err) console.error(err); console.log(result); })
QueryReqWrap {
  bindingName: 'queryA',
  callback: [Function],
  hostname: 'en',
  oncomplete: [Function: onresolve],
  ttl: true,
  search: true,
  domain:
   Domain {
     domain: null,
     _events:
      { removeListener: [Function: updateExceptionCapture],
        newListener: [Function: updateExceptionCapture],
        error: [Function: debugDomainError] },
     _eventsCount: 3,
     _maxListeners: undefined,
     members: [] },
  channel:
   ChannelWrap {
     domain:
      Domain {
        domain: null,
        _events: [Object],
        _eventsCount: 3,
        _maxListeners: undefined,
        members: [] } } }
> [ { address: '103.102.166.224', ttl: 228 } ]
```

I can verify this against `nslookup`:
```
$ nslookup "en"
Server:		10.11.1.20
Address:	10.11.1.20#53

Non-authoritative answer:
Name:	en.wikipedia.org
Address: 103.102.166.224
```

SirR4T replied:
I think I'm coming across a weird behaviour too, not sure if this is a bug though.

Steps to reproduce:
1. Start `./out/Release/node`, the build result of [this PR's branch](https://github.com/SirR4T/node/tree/fix19548-support-cares-search)
2. Require dns module.
3. launch this query for `"en"`: `dns.resolve4('en', {ttl: true, search: true}, (err, result) => { if (err) console.error(err); console.log(result); })`
4. Observe that `ESERVFAIL` error is thrown (this is expected).
5. In a different shell, update `/etc/resolv.conf` to have this line at the end: `search wikipedia.org`
6. Switch back to original shell, relaunch the same query
7. Observe that `ESERVFAIL` is still thrown.
8. Shutdown the repl and start again
9. Relaunch the query, and observe that it now works!

I expect that the `dns.resolve4()` query should work with the latest changes in the `/etc/resolv.conf` file, but I suspect it is getting cached somewhere, during startup / require.

SirR4T commented:
@silverwind : From [the docs](https://c-ares.haxx.se/docs.html), it seems as if the `ares_addrttl` and `ares_addr6ttl` structs are set in the `ares_parse_<type>_reply()` parsing functions only when `<type>` is `a` or `aaaa`, respectively. I do not think this is simply portable across other query types?

silverwind commented:
You're right, now that you mention it, I recall this as a c-ares limitation. Maybe make the `ttl` option only apply when `rrtype` is `A` and `AAAA`, if that's not too much of a hassle.

SirR4T commented:
we do support the `options` object only when using `dns.resolve4()` or `dns.resolve6()`, really. [The docs](https://nodejs.org/api/dns.html#dns_dns_resolve4_hostname_options_callback) reflect this too, as the `options` object or the `ttl` parameter isn't documented anywhere else...

Do you mean we should throw errors on receiving options object, for other `resolveXtype()`s? Currently, we silently ignore them as far as I can see:

```
> dns.resolveMx('nodejs.org', {ttl: true}, (err, result) => { if (err) console.error(err); console.log(result); })
QueryReqWrap {
  bindingName: 'queryMx',
  callback: [Function],
  hostname: 'nodejs.org',
  oncomplete: [Function: onresolve],
  ttl: true,
  domain:
   Domain {
     domain: null,
     _events: { error: [Function: debugDomainError] },
     _eventsCount: 1,
     _maxListeners: undefined,
     members: [] },
  channel: ChannelWrap {} }
> [ { exchange: 'aspmx2.googlemail.com', priority: 30 },
  { exchange: 'aspmx3.googlemail.com', priority: 30 },
  { exchange: 'aspmx.l.google.com', priority: 10 },
  { exchange: 'alt1.aspmx.l.google.com', priority: 20 },
  { exchange: 'alt2.aspmx.l.google.com', priority: 20 } ]

>
```
And I fear we could be breaking things by throwing, if this was already (erroneously) being used.

lundibundi commented on lib/internal/dns/promises.js:
```diff
@@ -180,8 +180,9 @@ function createResolverPromise(resolver, bindingName, hostname, ttl) {
+    req.search = isSearch;
 
-    const err = resolver._handle[bindingName](req, hostname);
+    const err = resolver._handle[bindingName](req, hostname, isSearch);
```
I may be misunderstanding something, but is it needed to pass an argument `isSearch` while it is already in `req` object and it can be extracted from there?

SirR4T replied:
I'm not sure either, how JSland calls into C++ land, but looking at [this block of code](https://github.com/nodejs/node/blob/master/src/cares_wrap.cc#L1806-L1812), I figured this is how I'd have to add an additional parameter, and it worked.

If you can show me an example of unwrapping properties on an object, in C++ land, I'm willing to test this approach out and update if needed.

SirR4T replied:
Besides, what would be the better approach? Unwrapping properties would implicitly assume that the property names are part of the API contract, whereas passing in parameters would make the signature explicit.

lundibundi replied:
Since on the c++ you already have an `Local<Object>` I think you can just 
```c++
Local<Value> search = req_wrap_obj->Get(env->isolate(), FIXED_ONE_BYTE_STRING(env->isolate(), "search"));
CHECK(search->IsBoolean());
bool is_search = (search.As<Boolean>)->Value();
```

Though, yeah, I'm not sure if this is a correct way of doing things (no need to change this yet). I just thought I'd mention it since it is both in the object and as an argument. Is it needed to set it in the `req` object at all (maybe just an argument will be enough)?

SirR4T replied:
not really needed anywhere else, I guess. Will fix this.

rvagg replied:
You need to be consistent, it's not done on the non-promises version. I think the only reason `ttl` is set on `req` is so that we can tack `ttl` on to the response object in `onresolve()`, i.e. to connect the input with the output processing rather than interacting with c-ares at all. So I think it'd be appropriate to not attach anything new to `req` since we don't need it after the query is sent.

SirR4T commented on src/cares_wrap.cc:
```diff
@@ -1806,10 +1825,14 @@ static void Query(const FunctionCallbackInfo<Value>& args) {
   Local<Object> req_wrap_obj = args[0].As<Object>();
   Local<String> string = args[1].As<String>();
-  Wrap* wrap = new Wrap(channel, req_wrap_obj);
+  bool is_search = args[2]->IsTrue();
```
Are these two interchangeable statements?
- `bool is_search = args[2]->IsTrue();`
- `bool is_search = args[2]->Value();`

Or is one of them preferred / idiomatic over the other?

addaleax replied:
`IsTrue()` should be fine.

addaleax commented:
@SirR4T I don’t know how to tackle the `/etc/resolv.conf` issue without upstream changes? Maybe they would be open to accepting something like a `CARES_RESOLV_CONF` environment variable that we could point to a file we choose…

SirR4T commented:
k, thanks. Raised [an issue upstream](https://github.com/c-ares/c-ares/issues/220), let's see.

addaleax commented:
@sirr4t Sounds good! Also, feel free to try their [mailing list](http://cool.haxx.se/mailman/listinfo/c-ares) – I think the maintainers are a tad more active there.

SirR4T commented:
This PR should probably be tagged as blocked, for now.

Blocked on:
- upstream accepting [this fix](https://github.com/c-ares/c-ares/pull/221),
- depending on timelines, releasing it,
- patch being backported / version upgraded in to master

addaleax commented:
@SirR4T I’m not sure … I guess you could cherry-pick it if you like? I don’t really know how frequently c-ares does releases, but it doesn’t seem to be too often?

SirR4T commented:
maybe @rvagg could weigh in? Asking since i see some recent activity from him, in deps/cares.

SirR4T commented:
actually, even after fixing the upstream issue I'm still not very clear how we would be able to inject mock `resolv.conf` files, while testing in JS land.

silverwind commented:
Any update on whether we can support this option on `.resolve` (specifically `Resolver.prototype.resolve`)? It should not be restriced to `A` and `AAAA` only.

SirR4T commented:
Hi @silverwind , all queries are now supported. Specifically:

- [x] resolveAny
- [x] resolve4
- [x] resolve6
- [x] resolveCname
- [x] resolveMx
- [x] resolveNs
- [x] resolveTxt
- [x] resolveSrv
- [x] resolvePtr
- [x] resolveNaptr
- [x] resolveSoa
- [x] reverse

apapirovski commented:
@SirR4T would you like to follow up on this? I believe the upstream issue has been resolved and an official version update released?

also /cc @addaleax

SirR4T commented:
Hi @apapirovski, I think I'm currently stuck trying to  point resolv.conf file in different test cases to different fixtures. Apart from that, if you're asking for updating the cares version, that could be handled as well. Do let me know if that needs a different PR or should be handled here itself

rvagg commented:
@SirR4T re your commits, have a look at doc/guides/contributing/pull-requests.md for commit message formatting. Also have a look in the git history, do a regex search for `deps.*cherry` to see how we do upstream cherry-pick commits, you'll need to be more verbose in how you've pulled that in.

FWIW I'm +1 on cherry-picking from c-ares, they release very sporadically and https://github.com/c-ares/c-ares/commit/7e6af8efbcd9b6a368d2d0d6a1226e164bc07f75 looks pretty safe.

addaleax commented:
@rvagg I just landed https://github.com/nodejs/node/pull/23854, no reason to cherry-pick anything :)

rvagg commented on deps/cares/src/ares_init.c:
```diff
@@ -1748,7 +1767,14 @@ static int init_by_resolv_conf(ares_channel channel)
 
-    fp = fopen(PATH_RESOLV_CONF, "r");
+    /* Support path for resolvconf filename set by ares_init_options */
+    if(channel->resolvconf_path) {
```
Node's C++ style has a space here (yes, confusing since c-ares has the all-squished-up style), run `make lint-cpp` and you should seen an error

rvagg commented:
OK, so I can't think of a clean way to properly test this either.

Unclean ways include:

 * Adding a magic environment variable that adjusts our `ares_options`, but then you get into the same discussion you had over with the c-ares folks about introducing magic for no user-facing reason. I can't come up with a good justification for that (in fact, adding more magic env vars opens up minor security concerns).
* If we could run `chroot` as non-root then that'd work, but alas we can't. We could add a test case that first checks for the existence of `fakechroot` on the system and then optionally runs the test if it exists. But then we'd have to remember to put it on as many systems as we can and we'd probably be limited to Linux (which isn't terrible but not ideal).
* Having a test case that parses `/etc/resolv.conf` and extracts any `search` property. If one exists, run a lookup for `www` and `www.${search value}` and compare. Unfortunately this depends on both having a `search` property and also having a `www.whatever` setup. I've checked a few of our CI hosts and none of our big infra providers give us a `search` to work with. Some hosts do, but they're minor, like some of the ARM machines. So we'd get some coverage but not a whole lot.
* Adjusting CI hosts via our Ansible setup scripts to insert something novel into every `/etc/resolv.conf`, maybe `search nodejs.org` and then we have full control and could do DNS lookups for whatever we wanted (make up a test value for every query type even!). The catch here is that resolv.conf is usually set by DHCP so they'd get overwritten each time the machine reboots. Setting resolv.conf to read-only is an option but would screw us if we had a necessary DNS change that came in via DHCP (although just using 1.1.1.1 would fix that potential problem I suppose). There's a bunch of other ways to override DHCP too, but every distro is different (see https://www.cyberciti.biz/faq/dhclient-etcresolvconf-hooks/ for the plethora of ways) and I don't even know what's possible on non-Linux!

Out of all of those I think the last option is most reasonable. Unfortunately that's going to require some major work on Ansible scripts which is not really a reasonable ask IMO.

Unless anyone else has any good ideas I think I'm inclined to say that we could just accept this without comprehensive tests. Some basic tests to make sure that passing `search` doesn't crash might have to be enough.

(also btw @addaleax I suspect that given all of the above that the c-ares update (or the floating patch) isn't really even helpful).

rvagg commented:
For @nodejs/tsc: there's a question here about whether the impracticality of coming up with workable test cases should dictate whether something can land. If we can't find a way to invoke the code path in question here in a reliable way within our testing infra, does that mean that we can't land this? I don't recall us having a discussion (lately anyway) about the burden of testability. Thoughts? (See my last comment here if you want details about potential paths, which are all sub-optimal).

Trott commented:
@nodejs/build on @rvagg's testing options above. 

I don't suppose the fact that we run `internet` tests daily on CI could somehow be leveraged, could it?

fhinkel commented:
I'm cleaning out a few old PRs 🧹. I'm closing this due to inactivity. Please re-open if needed!

strideynet commented:
I think it's a shame to let this PR close off based on an inability to effectively test it. Would there be any potential opposition to @rvagg's suggested solution of using ansible scripts to insert some dummy search domain onto the test hosts?
